src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlSerializer.cs

```diff
@@ -28,617 +28,816 @@
+                {
+                    if (!CanCreateValidUri(namespaceValue, UriKind.Absolute))
+                        throw LogReadException(LogMessages.IDX11111, SamlConstants.Elements.Action, SamlConstants.Attributes.Namespace, namespaceValue);
+                }
+                else
+                    namespaceValue = SamlConstants.DefaultActionNamespace;
 
-            reader.MoveToContent();
-            reader.ReadEndElement();
+                var action = reader.ReadElementContentAsString();
+                reader.MoveToContent();
+                return new SamlAction(action, new Uri(namespaceValue));
+            }
+            catch (Exception ex)
+            {
+                if (ex is SamlSecurityTokenReadException)
+                    throw;
 
-            return (string.IsNullOrEmpty(ns)) ? new SamlAction(action) : new SamlAction(action, ns);
+                throw LogReadException(LogMessages.IDX11112, ex, SamlConstants.Elements.Action, ex);
+            }
         }
 
+        /// <summary>
+        /// Reads the &lt;saml:Advice> element.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// The Advice element has an extensibility point to allow XML elements
+        /// from non-SAML namespaces to be included. By default, because the 
+        /// Advice may be ignored without affecting the semantics of the 
+        /// assertion, any such elements are ignored. To handle the processing
+        /// of those elements, override this method.
+        /// </para>
+        /// </remarks>
+        /// <param name="reader">A <see cref="XmlReader"/> positioned at a <see cref="SamlAdvice"/> element.</param>
+        /// <returns>A <see cref="SamlAdvice"/> instance.</returns>
         protected virtual SamlAdvice ReadAdvice(XmlDictionaryReader reader)
         {
-            if (reader == null)
-                throw LogHelper.LogArgumentNullException(nameof(reader));
-
-            var advice = new SamlAdvice();
+            XmlUtil.CheckReaderOnEntry(reader, SamlConstants.Elements.Advice, SamlConstants.Namespace);
 
-            // SAML Advice is an optional element and all its child elements are optional
-            // too. So we may have an empty saml:Advice element in the saml token.
-            if (reader.IsEmptyElement)
+            try
             {
-                // Just issue a read for the empty element.
+                var advice = new SamlAdvice();
+
+                // @xsi:type
+                XmlUtil.ValidateXsiType(reader, SamlConstants.Types.AdviceType, SamlConstants.Namespace);
+
+                // SAML Advice is an optional element and all its child elements are optional
+                // too. So we may have an empty saml:Advice element in the saml token.
+                if (reader.IsEmptyElement)
+                {
+                    // Just issue a read for the empty element.
+                    reader.MoveToContent();
```

Comment:
you always call reader.MoveToContent(), reader.Read()
move them out of the if clause and delete them below